### PR TITLE
Remove ros arguments

### PR DESCRIPTION
**Pull request by mjcarroll**
Addresses rclcpp component of ros2/rcl#219

Connects to ros2/rcl#219

### CODE REVIEW

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4127/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1208/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3410/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4234/)

File: rclcpp/include/rclcpp/utilities.hpp [hunk 1]
```diff
@@ -45,15 +46,40 @@ std::string to_string(T value)
...
 RCLCPP_PUBLIC
 void
-init(int argc, char * argv[]);
+init(int argc, char const * const argv[]);
+
+/// Remove ROS-specific arguments from argument vector.
+/**
+ * Some arguments may not have been intended as ROS arguments. This function
+ * populates a the aruments in a vector. Since the first argument is always assumed
+ * to be a process name, the vector will always contain the process name.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/utilities.hpp]: Same style comment here, one line per sentence.

File: rclcpp/include/rclcpp/utilities.hpp [hunk 2]
```diff
@@ -45,15 +46,40 @@ std::string to_string(T value)
...
+/**
+ * This does everything that init does, but additionally removes ROS-specific command line
+ * arguments
+ * \param[in] argc Number of arguments.
+ * \param[in] argv Argument vector.
+ * \return Members of the argument vector that are not ROS arguments.
+ */
+RCLCPP_PUBLIC
+std::vector<std::string>
+init_and_remove_ros_arguments(int argc, char const * const argv[]);
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/utilities.hpp]: We could just change the signature of `init` to always return the pruned list, and users could just ignore it.

File: rclcpp/src/rclcpp/utilities.cpp [hunk 3]
```diff
@@ -191,6 +191,54 @@ rclcpp::init(int argc, char * argv[])
...
+std::vector<std::string>
+rclcpp::remove_ros_arguments(int argc, char const * const argv[])
+{
+  rcl_allocator_t alloc = rcl_get_default_allocator();
+  rcl_arguments_t parsed_args;
+
+  if (rcl_parse_arguments(argc, argv, alloc, &parsed_args) != RCL_RET_OK) {
+    std::string msg = "failed to parse arguments: ";
+    msg += rcl_get_error_string_safe();
+    rcl_reset_error();
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: There is a function for this: https://github.com/ros2/rclcpp/blob/947e3f7e677df87db4e4b76bc13c0b8d217fbeb1/rclcpp/include/rclcpp/exceptions.hpp#L100-L119

File: rclcpp/src/rclcpp/utilities.cpp [hunk 4]
```diff
@@ -191,6 +191,54 @@ rclcpp::init(int argc, char * argv[])
 #endif
 }
 
+std::vector<std::string>
+rclcpp::remove_ros_arguments(int argc, char const * const argv[])
+{
+  rcl_allocator_t alloc = rcl_get_default_allocator();
+  rcl_arguments_t parsed_args;
+
+  if (rcl_parse_arguments(argc, argv, alloc, &parsed_args) != RCL_RET_OK) {
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: @sloretz was there going to be a function to get the already parsed arguments?

`rcl_init` already does this work and keeps the `rcl_arguments_t` around for later use. So we could use that here. The downside is that this would need `rcl_init` to have been called before, which might actually be a reason to leave this as-is.

Also, we'd need to ensure the `argc` and `argv` are the same ones given to the previously called `rcl_init`.

So with those two caveat's maybe this is the best approach.

File: rclcpp/src/rclcpp/utilities.cpp (hunk 3 above)
Comment (wjwwood) [reply rclcpp/src/rclcpp/utilities.cpp]: There are other opportunities to use this below too.

File: rclcpp/src/rclcpp/utilities.cpp [hunk 5]
```diff
@@ -191,6 +191,54 @@ rclcpp::init(int argc, char * argv[])
...
+      &parsed_args,
+      alloc,
+      &nonros_argc,
+      &nonros_argv) != RCL_RET_OK)
+  {
+    std::string msg = "failed to remove ROS arguments: ";
+    msg += rcl_get_error_string_safe();
+    rcl_reset_error();
+    if (NULL != nonros_argv) {
+      alloc.deallocate(nonros_argv, alloc.state);
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: Same question as on the other pr, I would expect this to not be required. I would expect `rcl_remove_ros_arguments` to clean up memory allocations in an error condition, so that it is not required here.

File: rclcpp/src/rclcpp/utilities.cpp [hunk 6]
```diff
@@ -191,6 +191,54 @@ rclcpp::init(int argc, char * argv[])
...
+    }
+    throw std::runtime_error(msg);
+  }
+
+  std::vector<std::string> return_arguments;
+  return_arguments.resize(nonros_argc);
+
+  for (int ii = 0; ii < nonros_argc; ++ii) {
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: I'd use `size_t` rather than `int`, unless you had a specific reason here.

File: rclcpp/test/test_utilities.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "__ns:=/fiz/buz", "--foo=bar", "--baz"};
+  int argc = sizeof(argv) / sizeof(const char *);
+  rclcpp::init(argc, argv);
+  auto args = rclcpp::remove_ros_arguments(argc, argv);
+
+  ASSERT_EQ(4u, args.size());
+  ASSERT_EQ(std::string{"process_name"}, args[0]);
+  ASSERT_EQ(std::string{"-d"}, args[1]);
+  ASSERT_EQ(std::string{"--foo=bar"}, args[2]);
+  ASSERT_EQ(std::string{"--baz"}, args[3]);
```
Comment (wjwwood) [inline rclcpp/test/test_utilities.cpp]: It would be good to see a test for `init_and_remove_ros_arguments` also or instead.

Also, I'd like to see tests for corner cases, like `argc = 0` and/or `argv = nullptr`.

File: rclcpp/src/rclcpp/utilities.cpp (hunk 4 above)
Comment (sloretz) [reply rclcpp/src/rclcpp/utilities.cpp]: @wjwwood You mean a function that gets the global parsed arguments? There is [`rcl_get_global_arguments`](https://github.com/ros2/rcl/blob/e48a445f46f1da21ff0c4bf3d52bec9120ab7b57/rcl/include/rcl/arguments.h#L181). Both your caveats apply.

File: rclcpp/src/rclcpp/utilities.cpp (hunk 6 above)
Comment (sloretz) [reply rclcpp/src/rclcpp/utilities.cpp]: @wjwwood if it is `size_t` will windows warn about signed/unsigned comparison?

File: rclcpp/src/rclcpp/utilities.cpp (hunk 4 above)
Comment (wjwwood) [reply rclcpp/src/rclcpp/utilities.cpp]: Oh yeah, forgot you ended up adding that. Any opinion on how this function should work, i.e. should it reused the result of init or do it again itself to avoid the corner cases?

File: rclcpp/src/rclcpp/utilities.cpp (hunk 6 above)
Comment (wjwwood) [reply rclcpp/src/rclcpp/utilities.cpp]: That's a fair point, it might.

File: rclcpp/src/rclcpp/utilities.cpp (hunk 4 above)
Comment (sloretz) [reply rclcpp/src/rclcpp/utilities.cpp]: As a standalone function with `argc` and `argv` passed in +1 to parsing the args again to prevent a mismatch between the arguments and the global state.

A non-public `rclcpp` function that made an `std::vector` of non-ros arguments given `argc`, `argv`, and `rcl_arguments_t` could be used by both `remove_ros_arguments` and  `init_and_remove_ros_arguments` to avoid the double-parsing. However, inefficiency in `init..()` is a one-time cost in most use cases so it may not be worth the added work.

Comment (wjwwood) [reply rclcpp/src/rclcpp/utilities.cpp]: Sounds good to me too. @mjcarroll I'll leave it up to you to either use the common function or not as @sloretz described.

Comment (mjcarroll) [reply rclcpp/src/rclcpp/utilities.cpp]: This issue would be OBE if we collapse init into a single signature that always returns a std::vector.

File: rclcpp/src/rclcpp/utilities.cpp (hunk 6 above)
Comment (mjcarroll) [reply rclcpp/src/rclcpp/utilities.cpp]: Also MISRA, but that's not as important here.

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4128/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1209/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3411/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4235/)

File: rclcpp/src/rclcpp/utilities.cpp [hunk 8]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: minimize vertical whitespace: https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

File: rclcpp/src/rclcpp/utilities.cpp [hunk 9]
```diff
@@ -191,6 +193,57 @@ rclcpp::init(int argc, char * argv[])
...
+{
+  rclcpp::init(argc, argv);
+  return rclcpp::remove_ros_arguments(argc, argv);
+}
+
+std::vector<std::string>
+rclcpp::remove_ros_arguments(int argc, char const * const argv[])
+{
+  rcl_allocator_t alloc = rcl_get_default_allocator();
+  rcl_arguments_t parsed_args;
```
Comment (sloretz) [inline rclcpp/src/rclcpp/utilities.cpp]: Need to zero initialize this before calling `rcl_get_zero_initialized_arguments()`. This could be the source of the random segfault.

Comment (sloretz) [reply rclcpp/src/rclcpp/utilities.cpp]: Possible cause of #458
